azure-pipelines.yml

```diff
@@ -14,281 +16,19 @@ trigger:
       - v2.0.0-beta.*
 
 jobs:
-  - job: Security
-    displayName: 'Security Tasks'
-    pool:
-      vmImage: 'windows-latest'
-    steps:
-      - task: ESLint@1
-        displayName: 'Run ESLint'
-        inputs:
-          Configuration: 'required'
-          TargetType: 'eslint'
-          TargetsESLint: '$(Build.SourcesDirectory)/**/*.{js,jsx,ts,tsx}'
-          ErrorLevel: 'warn'
-
-      - task: CredScan@3
-        displayName: 'Run Credential Scanner'
-        condition: succeededOrFailed()
-        inputs:
-          debugMode: false
-          suppressionsFile: '.sdl\CredScanSuppressions.json'
-
-      - task: PublishSecurityAnalysisLogs@3
-        inputs:
-          ArtifactName: 'CodeAnalysisLogs'
-          ArtifactType: 'M365'
-          AllTools: true
-          ToolLogsNotFoundAction: 'Standard'
-        condition: succeededOrFailed()
-        displayName: 'Publish Guardian Artifacts 2'
-
-        # Retaining artifacts using Arrow service (https://aka.ms/m365sdlonboarding​)
-      - task: AssetRetention@3
-        inputs:
-          ArrowServiceConnection: '$(ArrowConnection)'
-          AssetGroupName: '$(System.TeamProject)_$(Build.DefinitionName)'
-          AssetNumber: '$(Build.BuildId)'
-          IsShipped: false
-          DropsToRetain: 'CodeAnalysisLogs'
-        condition: and(
-          startsWith(variables['Build.SourceBranch'], 'refs/tags/v2.0.0-beta'),
-          in(variables['Agent.JobStatus'], 'Succeeded', 'SucceededWithIssues'),
-          ne(variables['Build.Reason'], 'PullRequest'),
-          ne(variables['Build.Reason'], 'Manual')
-          )
-        displayName: 'Artifact Retention(Arrow)'
-
-      - task: PostAnalysis@2
-        condition: succeededOrFailed()
-        displayName: 'Guardian Break'
-        inputs:
-          GdnBreakPolicyMinSev: Warning
-          GdnBreakAllTools: true
-          GdnBreakGdnToolCredScan: true
-          GdnBreakGdnToolCredScanSeverity: Warning
-          GdnBreakGdnToolESLint: true
-          GdnBreakGdnToolESLintSeverity: Warning
-          GdnBreakPolicy: M365
-
-      - task: ComponentGovernanceComponentDetection@0
-        condition: and(
-          eq(variables['Build.Reason'], 'PullRequest'),
-          eq(variables['System.PullRequest.TargetBranch'], '2.0-preview'))
-        inputs:
-          scanType: 'LogOnly'
-          verbosity: 'Verbose'
-          alertWarningLevel: 'High'
-          failOnAlert: true
-        displayName: 'Component Governance (Pull Request)'
+  - template: tools/yaml/security.yml
```

Comment:
Do you think there is value in storing the yaml templates in a subdirectory called templates? Relatedly, is there a problem if any of the template yaml files (like security.yml) is used directly or in a certain way, and if so, would we be more likely to avoid using it in a bad way if it were more clearly obvious that it's a template? 

I'm new to templates, so trying to figure out if we should be more obvious about their use.

Reply:
That is a great question!

The main reason I'm doing this is simply syntactical (we had an earlier discussion about being able to merge into one yaml file without a cost of readability since our yaml file is like 500 lines now). Templates solve this, at the risk of then being reused incorrectly if we ever need to make more yaml pipelines and reuse templates. 

I don't know if we'll really ever need to make a new yaml pipeline, so it really depends on what another pipeline is trying to do if it can then reuse the steps here. I think if that ever happens, based on new pipeline requirements, we can edit the templates to then fit both pipelines. (ie security.yml and build-test-publish.yml are jobs files, whereas build-app-host is a steps file, and you can see how they are used different in the azure-pipelines.yml file now)

I doubt the 1st two templates I mentioned are ever going to be reused. The only reason I made them templates is to make the main file easier to read. I am ok with keeping them inside however.
The nice thing about the build-app-host.yml however is it shortens all the copied steps for building the app host before running e2e tests. So the template build-app-host does all the checking out and building both repos, and then we just add a script afterwards in each job to run the specific e2e test.